Write APICompat suppressions when baseline suppressions aren't empty

ViktorHofer opened the pull request:
When there aren't any suppressions but baseline suppressions exist, write to the suppression file to make it empty.

Pass information back to the caller so that even if an empty collection of suppression are written to the file, still log a message to indicate that the suppression file got updated.

ViktorHofer committed:
Write APICompat suppressions when baseline suppressions aren't empty

When there aren't any suppressions but baseline suppressions exist,
write to the suppression file to make it empty.

Pass information back to the caller so that even if an empty collection
of suppression are written to the file, still log a message to indicate
that the suppression file got updated.

ViktorHofer commented:
@ericstj when the /p:GenerateAPICompatSuppressionFile=true` flag is passed in, the baseline suppression file not being empty and there aren't actual suppressions, should we write an empty suppressions file or not write at all and try to delete if it exists?

With this change, we update the file and make it empty:

```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- https://learn.microsoft.com/dotnet/fundamentals/package-validation/diagnostic-ids -->
<Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" />
```

ericstj commented on src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:
```diff
@@ -104,19 +104,22 @@ public bool IsErrorSuppressed(Suppression error)
             // If unnecessary suppressions should be preserved in the suppression file, union the
             // baseline suppressions with the set of actual suppressions. Duplicates are ignored.
-            HashSet<Suppression> suppressionsToSerialize = new(_suppressions);
+            HashSet<Suppression> suppressionsToSerialize = [.. _suppressions];
```
I'm not so sure this is better syntax.  It doesn't call the constructor that takes a collection, so it initializes with default capacity then Add's N times.  This can result in reallocation and copies.  I think you did this because an analyzer told you to -- maybe we should also follow up with that analyzer.

ericstj replied:
I opened https://github.com/dotnet/roslyn/issues/75981.

ViktorHofer replied:
Yeah, analyzer in this case.

ericstj commented:
> @ericstj when the `/p:GenerateAPICompatSuppressionFile=true` flag is passed in, the baseline suppression file not being empty but there aren't actual suppressions, should we write an empty suppressions file or not write at all and try to delete if it exists?
> 
> With this change, we update the file and make it empty:
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> <!-- https://learn.microsoft.com/dotnet/fundamentals/package-validation/diagnostic-ids -->
> <Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" />
> ```

I think in our case we'd prefer to delete the file since we conditionally set it based on existence -- would that work in general?  Suppose someone specified the file name - if we deleted that file would we handle a specified path to a non-existent file on run?  Maybe not.  I guess emitting an empty file is OK.  If folks want to delete it, then can notice the empty file and delete it.

ViktorHofer commented:
> Suppose someone specified the file name - if we deleted that file would we handle a specified path to a non-existent file on run? Maybe not. I guess emitting an empty file is OK. If folks want to delete it, then can notice the empty file and delete it.

The suppression engine file loader would throw a FileNotFoundException: https://github.com/dotnet/sdk/blob/5ef57c869107648581147d67ab6b137ed55e04f2/src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs#L45

I lean towards deleting the empty suppression file. The switch name might confusing here though as it implies something getting generated, not deleted (`GenerateAPICompatSuppressionFile`).

ericstj commented:
I'm fine either way, just test the behavior works in and doesn't introduce new errors for folks.

ViktorHofer committed:
Update SuppressionEngine.cs

ericstj commented on test/Microsoft.DotNet.ApiCompatibility.Tests/Logging/SuppressionEngineTests.cs:
```diff
@@ -60,8 +60,9 @@ public void SuppressionEngine_WriteSuppressionsToFile_SuppressionsRoundTrip()
-            IReadOnlyCollection<Suppression> writtenSuppressions = suppressionEngine.WriteSuppressionsToFile(filePath, preserveUnnecessarySuppressions: true);
+            var (suppressionFileUpdated, writtenSuppressions) = suppressionEngine.WriteSuppressionsToFile(filePath, preserveUnnecessarySuppressions: true);
 
+            Assert.True(suppressionFileUpdated);
```
Should we add a test case for the scenario where we had baselines and they are no longer needed -- ensure that the file is emptied and true is returned?

ViktorHofer replied:
I just noticed that we already have one that validates that: `SuppressionEngine_WriteSuppressionsToFile_ReturnsEmptyWithAllUnnecessarySuppressions`

ericstj reviewed (approved):
change looks ok, please consider adding a test for this scenario.

ViktorHofer commented:
Yes, I will add a test for that scenario before merging.

ViktorHofer committed:
Update SuppressionEnigneTests

The SuppressionEngine_WriteSuppressionsToFile_ReturnsEmptyWithAllUnnecessarySuppressions
test already validates the new behavior.

ViktorHofer committed:
Update SuppressionEngineTests.cs
